Course versioning redirect feedback

maddiedierker opened the pull request:
### What it does
- Refactors helpers for dismissing version warnings/dialogs into a helper in utils
- Stores a session cookie when a user dismisses a version dialog
- We want the dismissal of a script version dialog to count for all scripts in that course, so store the course name in the session cookie wherever possible

Implements the following feedback (from [this doc](https://docs.google.com/document/d/1DfYNku2iC5QuBd9EX2OxDY0nhLbCyfhGJPmNuzXrMO8/edit?usp=sharing)):
- Make sure that the above “you’re not in the right place” dialog doesn’t show up again when I dig further into the course / unit. It is currently (needs a session cookie?)
- (P2) Go to old version of CSP (2017) and saw the version warning (correct), but when I clicked on a unit, still saw the warning on the unit page. Expected: The warning should be hidden at this point because I didn’t try to go to the old version.

davidsbailey commented:
👀 looking...

davidsbailey commented:
for other course version warnings, we remember forever that a user has dismissed a warning by storing a record of it in the user_scripts table: https://github.com/code-dot-org/code-dot-org/blob/a11a00edae28f881436e89d7ed2db4e283dcc146/dashboard/app/models/course.rb#L436-L444

Are we pretty sure we need two separate implementations for this? Let's not block this PR on it, but if we want to unify these implementations then let's get a work item into the backlog to track this as tech debt.

maddiedierker commented:
@davidsbailey good question. poorva and i specifically talked about having these warnings/dialogs be session-based, so i think it's worth having a conversation about whether which implementation is preferred for all warnings/notifications/dialogs about version warnings. 

@poorvasingal thoughts? should all of our versioning warnings be session-based, or should we remember them forever? is there a reason for any of them to be implemented differently?

davidsbailey commented on apps/src/util/dismissVersionRedirect.js:
```diff
@@ -0,0 +1,38 @@
+const dismissVersionRedirect = module.exports;
```
This looks like the old way of exporting things. Instead, you could `export const DISMISSED_REDIRECT_WARNINGS_SESSION_KEY ...`, and then in the other file, `import * as dismissVersionRedirect from ...`. unless there's a reason that's not working.

davidsbailey replied:
looks like the constant I used as an example isn't exported. so a better example might be 
`export const dismissedRedirect = function(...) {` or `export function dismissedRedirect(...) {`

davidsbailey replied:
> then in the other file, `import * as dismissVersionRedirect from ...`

Actually, it seems like it would be even better to import the items individually, as 
```js
import {dismissedRedirectWarning, dismissedRedirectDialog, ...} from ...`
```
sorry to go around so many times on this 😳

davidsbailey commented on apps/src/util/dismissVersionRedirect.js:
```diff
@@ -0,0 +1,38 @@
+
+dismissVersionRedirect.dismissedRedirect = function (sessionKey, name) {
+  const dismissedRedirects = sessionStorage.getItem(sessionKey);
+  return (dismissedRedirects || '').includes(name);
```
it looks like this could accidentally match substrings. it would be better to do `return (dismissedRedirects || '').split(',').includes(name);`

poorvasingal commented:
We made these session based because we wanted to do our best to keep preventing students from accessing old scripts. So I was nervous if they choose the option to go to the old version once, they wouldn't ever be warned about it again in the future. 

Can we log to firehose whenever someone sees that dialog? I'd like to track user id, what link they were trying to go to, and which option they chose. That way we can see if users keep hitting this dialog over and over again. That'll probably help us figure out what the right long term approach is here.

davidsbailey commented:
Logging to Firehose when this warning dialog is shown sounds good to me.

maddiedierker commented:
thanks, @davidsbailey and @poorvasingal! i'll implement logging to firehose in a separate PR.

@davidsbailey i think i addressed all of your comments, so please take a look! 😃
